### DISCUSSION TITLE
[Q&A] Any plans about developing collection-level segment maxSize?

### DISCUSSION THREAD

**Discussion by SeunghyunYoo** [milvus-io/milvus#28633] [Q&A and General discussion]
Do you guys have any plan for developing collection-level segment maxSize as collection properties?

Segment maxSize affect compaction plan, so compaction occurs too frequently in small collection which have smaller size than the maxSize.

If i run milvus cluster only for similar size of collection, then maybe i can handle this situation by tuning parameters like `dataCoord.segment.maxSize`, `compactableProportion`, `SegmentExpansionRate`.

However if i run milvus cluster for several size of collection, it's hard to tuning cluster-level parameters.

In some situation, maybe it can be solved by creating new collection and switching it.

However, it is also hard to solve if i upsert new data every 1 hour for some reason.
In my case, i need frequently updatable index for recommendation system without search performance loss from brute-force search (growing segment or sealed segment which have no index due to compaction).

Beside, to run milvus cluster for for several size of collection which needs frequently updatable index, it seems like more hard.

---

So, I want to know do you guys any plan for collection-level segment maxSize or there are any difficulties and side effects to develop it.

Plus, I want to know any milvus operation experience to handle that kinds of situation.

thanks.

**Comment by yhmo** [comment]
Change segment maxSize is not a good solution to solve "search performance loss from brute-force search on growing segment or sealed segment which have no index due to compaction".
Compaction is single-direction operation, small segments can be merged to large segments, but large segments cannot be split to small segments. If the segment.maxSize is 1GB, and you change it to 512MB, nothing changes to the segments.

So far as I know, auto index for growing segments might be a better solution for the problem you mentioned. We have internal prototype and testing, seems there are some technical issues and this feature was not published.

**Reply by SeunghyunYoo** [reply]
Suppose that data is small but `segment.maxSize` is big.

Then, whenever data insert, the segment will be merged with new small segment.
In this time, newly segment is loaded which have no index until the segment is loaded which have index after building index done. (1)
Then search request to the newly loaded segment become the brute force.

Isn't it?
I worried this situation, so i tried to tune `dataCoord.segment.maxSize` or `SegmentExpansionRate` (1.25 -> 1)

If i misunderstand the (1) situation, please tell me.

thanks

---

> So far as I know, auto index for growing segments might be a better solution for the problem you mentioned. We have internal prototype and testing, seems there are some technical issues and this feature was not published.

Plus, I also look forward that kinds of feature.
If it is possible, it will be very great :)

**Reply by yhmo** [reply]
Since the growing segments are in memory, they are already loaded state, they have no index so the search is brute force.
Once a growing segment size exceeds a threshold(segment.sealProportion * segment.maxSize), milvus turns it to be "sealed segment". At this moment, this sealed segment has no index, it still be kept in query node's memory to do brute-force search. Meanwhile, index node begin to build index for this sealed segment.
After index is finished, query node will load its index into memory to replace the sealed segment. Search is index search from now on.

The seal threshold is **segment.sealProportion * segment.maxSize**, if you increase the value of segment.maxSize, growing segment size also increases, search performance gets worse when the size is near to the threshold. The search performance curve is like big waves:
```
    /|     /|     /|
  /  |   /  |   /  |
/    |_/    |_/    |
```

If you decrease the value of segment.maxSize, sealed segment is smaller, the number of segment increases. The search performance curve is like small waves:
```
  /|   /|   /|
/  |_/  |_/  |
```

**Reply by SeunghyunYoo** [reply]
I also worry about performance drop from growing segment.
But, it could be handled using `ignore_growing`

Plus, as you mentioned before, it could also be useful if i reduce `segment.maxSize`
(But, maybe there is side effect on big collection)

---

Except growing segment, I'm worry about following situation.

a. There is 1 `segmentA` loaded (smaller size of segment than `segment.maxSize`)

b. Insert new data to this collection

c. after times go, the growing segment become sealed `segmentB1`, `segmentC1`.
- in this time, there is no index on `segmentB1`, `segmentC1`
- **so, brute force period on `segmentB1`**

d. after building index, `segmentB2`,  `segmentC2` become loaded 
- `segmentB2` is indexed version of `segmentB1`
- no brute force period

e. compaction triggered on `segmentA`, `segmentB2`, `segmentC2`
- `segmentA` + `segmentB2` + `segmentC2`-> `segmentD1`
- suppose, `segmentA` + `segmentB2` + `segmentC2` is smaller than `segment.maxSize`

f. compaction done and `segmentD1` become loaded
- `segmentD1` has no index
- so, brute force period on `segmentD1`
- but, the collection has only 1 `segmentD1` in this time.
- **very pool search performance**

g. after building index, `segmentD2` become loaded
- `segmentD2` is indexed version of `segmentD1`
- no brute force period

Is the process right?

**Reply by yhmo** [reply]
In 2.3.x
- If a segment has no index, it cannot be compacted with others
- always try to load segment's index as much as possible

So, 
a. There is 1 segmentA1 loaded (smaller size of segment than segment.maxSize)
If it has no index, index node begins to build index for it. 
If its index is finished, the index segmentA2 is loaded to replace the segmentA1
segmentA2 is indexed version of segmentA1

b. Insert new data to this collection

c. after times go, the growing segment become sealed segmentB1, segmentC1.
in this time, there is no index on segmentB1, segmentC1
so, brute force period on segmentB1
They cannot be compacted since they have no index.

d. after building index, segmentB2, segmentC2 become loaded
segmentB2 is indexed version of segmentB1
no brute force period

e. compaction triggered on segmentA2, segmentB2, segmentC2
segmentA2 + segmentB2 + segmentC2-> segmentD1
suppose, segmentA2 + segmentB2 + segmentC2 is smaller than segment.maxSize

f. compaction done and segmentD1 generated
segmentD1 has no index, the segmentA2, segmentB2, segmentC2 are kept in memory to do index search
index node begins to build index for segmentD1
no brute force period, search performance is no change

g. after building index, segmentD2 become loaded
segmentD2 is indexed version of segmentD1
segmentD2 replaces the segmentA2, segmentB2, segmentC2
no brute force period

**Reply by SeunghyunYoo** [reply]
> f. compaction done and segmentD1 generated
segmentD1 has no index, the segmentA2, segmentB2, segmentC2 are kept in memory to do index search
index node begins to build index for segmentD1
no brute force period, search performance is no change

But in my test case, 
it seems like that `the segmentA2, segmentB2, segmentC2` are replaced to `segmentD1` before segmentD1.

Maybe I misunderstand the situation the index load fail at (d) due to `the context cancel issue` that i already report.
- https://github.com/milvus-io/milvus/discussions/28374#discussioncomment-7575807

**Reply by SeunghyunYoo** [reply]
By the way, To reduce search performance drop, it it better that after index build, sealed segment load to querynode? (at c)

Is there any other issue when sealed segment is loaded after index build

**Reply by yhmo** [reply]
The behavior of compaction/index/loading is always changing, it depends on your milvus version. I don't know all the details of the implementation of all the versions. I just describe the current major behavior in v2.3.x. There might be some other factors affect the behavior, such as small segments that row count of less than 1024, and upsert/delete actions also affect the behavior. And there might be some bugs that the behavior does not follow the expectation.

**Comment by xiaofan-luan** [comment]
Hi Seunghyun,
Totally understand your concern and we believe collection level configs is reasonable.

Tuning compaction algorithm is tricky, we don't actaully recommned to tune max segment size(it should be only related with your pod size, for 8GB-16GB memory pod we recommend to use 512M, for larger pod size we recommend to use 1G-2G)

On the other side, to tune flush segment size datacoord.segemnt.sealProportion is what you are looking for. With smaller sealProportion size you will be end up with flush more frequent.

Also, to be mentioned, on milvus 2.3.3 we fully supported growing segment index. With growing segment index, growing data will be indexed using IVF index(Soon we will support ScaNN). we already see 5 times acceleration on growing datas. Check if this can solve your problem

**Reply by SeunghyunYoo** [reply]
thanks, xiaofan

I tried to think up the way to handle the situation using `datacoord.segemnt.sealProportion`

However, I think it can't handle the compaction scenario as mentioned above (https://github.com/milvus-io/milvus/discussions/28633#discussioncomment-7637680)

Segment become sealed segment fast and index build also with `low sealProportion`. 
However, small sized segment become compaction candidates more easily, so the situation will happen more frequently i think.
 
To handle this issue, the merged segment have to be replaced (loaded) after index building.
Is it possible now?

Plus, I want to know exact segment load process. Could you tell me corresponding the code?

---

> Also, to be mentioned, on milvus 2.3.3 we fully supported growing segment index. With growing segment index, growing data will be indexed using IVF index(Soon we will support ScaNN). we already see 5 times acceleration on growing datas. Check if this can solve your problem

By the way, it's good news fully supported growing segment index on milvus 2.3.3 and ScaNN will be

I'll check if the feature can handle performance drop in the frequently data insert situation and share the results.

I have more questions about this feature.
After index is built on growing segment, when the growing segment become sealed, there is no additional index building process on the sealed segment?

**Comment by xiaofan-luan** [comment]
without compaction on smaller segments, you may not end of with a stable performance.

Yes, after flush/compaction happened, you will need to rebuild index. because the index built on growing segment is not as efficient as on sealed segment
